File: starters/node/src/transport.ts
import WebSocket, { WebSocketServer } from 'ws';

export enum TransportType {
  WebSocket = 'websocket',
  Quic = 'quic',
  Grpc = 'grpc',
}

export interface Conn {
  readMessage(): Promise<Buffer>;
  writeMessage(data: Buffer): Promise<void>;
  close(): Promise<void>;
  transportType(): TransportType;
  remoteAddr(): string;
}

export class WsServerConn implements Conn {
  private queue: Buffer[] = [];
  private waiters: Array<(data: Buffer) => void> = [];
  private closed = false;

  constructor(
    private ws: WebSocket,
    private addr: string,
  ) {
    ws.binaryType = 'nodebuffer';
    ws.on('message', (data: Buffer) => {
      if (this.waiters.length > 0) {
        this.waiters.shift()!(data);
      } else {
        this.queue.push(data);
      }
    });
    ws.on('close', () => {
      this.closed = true;
      for (const w of this.waiters) {
        w(Buffer.alloc(0));
      }
      this.waiters = [];
    });
  }

  readMessage(): Promise<Buffer> {
    if (this.queue.length > 0) {
      return Promise.resolve(this.queue.shift()!);
    }
    if (this.closed) {
      return Promise.reject(new Error('connection closed'));
    }
    return new Promise((resolve) => {
      this.waiters.push(resolve);
    });
  }

  writeMessage(data: Buffer): Promise<void> {
    return new Promise((resolve, reject) => {
      this.ws.send(data, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  close(): Promise<void> {
    this.ws.close();
    return Promise.resolve();
  }

  transportType(): TransportType {
    return TransportType.WebSocket;
  }

  remoteAddr(): string {
    return this.addr;
  }
}

export class WsClientConn implements Conn {
  private queue: Buffer[] = [];
  private waiters: Array<(data: Buffer) => void> = [];
  private closed = false;

  constructor(private ws: WebSocket) {
    ws.binaryType = 'nodebuffer';
    ws.on('message', (data: Buffer) => {
      if (this.waiters.length > 0) {
        this.waiters.shift()!(data);
      } else {
        this.queue.push(data);
      }
    });
    ws.on('close', () => {
      this.closed = true;
      for (const w of this.waiters) {
        w(Buffer.alloc(0));
      }
      this.waiters = [];
    });
  }

  static connect(url: string): Promise<WsClientConn> {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(url);
      ws.on('open', () => resolve(new WsClientConn(ws)));
      ws.on('error', reject);
    });
  }

  readMessage(): Promise<Buffer> {
    if (this.queue.length > 0) {
      return Promise.resolve(this.queue.shift()!);
    }
    if (this.closed) {
      return Promise.reject(new Error('connection closed'));
    }
    return new Promise((resolve) => {
      this.waiters.push(resolve);
    });
  }

  writeMessage(data: Buffer): Promise<void> {
    return new Promise((resolve, reject) => {
      this.ws.send(data, (err) => {
        if (err) reject(err);
        else resolve();
      });
    });
  }

  close(): Promise<void> {
    this.ws.close();
    return Promise.resolve();
  }

  transportType(): TransportType {
    return TransportType.WebSocket;
  }

  remoteAddr(): string {
    return this.ws.url ?? 'unknown';
  }
}

File: starters/rust/fproto-core/src/transport.rs
use std::fmt;
use thiserror::Error;
use tokio::net::TcpStream;
use tokio_tungstenite::{
    connect_async, accept_async,
    tungstenite::Message,
    MaybeTlsStream, WebSocketStream,
};
use futures_util::{SinkExt, StreamExt};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TransportType {
    WebSocket,
    Quic,
    Grpc,
}

impl fmt::Display for TransportType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TransportType::WebSocket => write!(f, "websocket"),
            TransportType::Quic => write!(f, "quic"),
            TransportType::Grpc => write!(f, "grpc"),
        }
    }
}

#[derive(Error, Debug)]
pub enum TransportError {
    #[error("websocket error: {0}")]
    WebSocket(#[from] tokio_tungstenite::tungstenite::Error),
    #[error("connection closed")]
    Closed,
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
}

pub struct WsServerConn {
    ws: WebSocketStream<TcpStream>,
    addr: String,
}

impl WsServerConn {
    pub fn new(ws: WebSocketStream<TcpStream>, addr: String) -> Self {
        Self { ws, addr }
    }

    pub async fn read_message(&mut self) -> Result<Vec<u8>, TransportError> {
        match self.ws.next().await {
            Some(Ok(Message::Binary(data))) => Ok(data),
            Some(Ok(Message::Close(_))) | None => Err(TransportError::Closed),
            Some(Ok(_)) => self.read_message().await,
            Some(Err(e)) => Err(TransportError::WebSocket(e)),
        }
    }

    pub async fn write_message(&mut self, data: &[u8]) -> Result<(), TransportError> {
        self.ws.send(Message::Binary(data.to_vec().into())).await?;
        Ok(())
    }

    pub async fn close(&mut self) -> Result<(), TransportError> {
        self.ws.close(None).await?;
        Ok(())
    }

    pub fn transport_type(&self) -> TransportType {
        TransportType::WebSocket
    }

    pub fn remote_addr(&self) -> &str {
        &self.addr
    }
}

pub struct WsClientConn {
    ws: WebSocketStream<MaybeTlsStream<TcpStream>>,
    addr: String,
}

impl WsClientConn {
    pub async fn connect(url: &str) -> Result<Self, TransportError> {
        let (ws, _) = connect_async(url).await?;
        Ok(Self {
            ws,
            addr: url.to_string(),
        })
    }

    pub async fn read_message(&mut self) -> Result<Vec<u8>, TransportError> {
        match self.ws.next().await {
            Some(Ok(Message::Binary(data))) => Ok(data),
            Some(Ok(Message::Close(_))) | None => Err(TransportError::Closed),
            Some(Ok(_)) => self.read_message().await,
            Some(Err(e)) => Err(TransportError::WebSocket(e)),
        }
    }

    pub async fn write_message(&mut self, data: &[u8]) -> Result<(), TransportError> {
        self.ws.send(Message::Binary(data.to_vec().into())).await?;
        Ok(())
    }

    pub async fn close(&mut self) -> Result<(), TransportError> {
        self.ws.close(None).await?;
        Ok(())
    }

    pub fn transport_type(&self) -> TransportType {
        TransportType::WebSocket
    }

    pub fn remote_addr(&self) -> &str {
        &self.addr
    }
}

pub async fn accept_ws(stream: TcpStream, addr: String) -> Result<WsServerConn, TransportError> {
    let ws = accept_async(stream).await?;
    Ok(WsServerConn::new(ws, addr))
}
